internal/pkg/service/stream/storage/level/local/writer/writer.go

```diff
@@ -1,42 +1,360 @@
+	}
+
+	// Measure size of compressed data
+	{
+		backupPath := filepath.Join(dirPath, CompressedSizeFile)
+		_, err = w.chain.PrependWriterOrErr(func(writer io.Writer) (out io.Writer, err error) {
+			w.compressedMeter, err = size.NewMeterWithBackupFile(ctx, clk, logger, writer, backupPath, cfg.Statistics.DiskSyncInterval)
+			return w.compressedMeter, err
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Add compression if enabled
+	{
+		if compConfig := slice.LocalStorage.Compression; compConfig.Type != compression.TypeNone {
+			// Add compression writer
+			_, err = w.chain.PrependWriterOrErr(func(writer io.Writer) (io.Writer, error) {
+				return compressionWriter.New(writer, compConfig)
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			// Add a buffer before compression writer, if it is not included in the writer itself
+			if cfg.InputBuffer > 0 && !compConfig.HasWriterInputBuffer() {
+				w.chain.PrependWriter(func(w io.Writer) io.Writer {
+					return limitbuffer.New(w, int(cfg.InputBuffer.Bytes()))
+				})
+			}
+
+			// Measure size of uncompressed CSV data
+			backupPath := filepath.Join(dirPath, UncompressedSizeFile)
+			_, err = w.chain.PrependWriterOrErr(func(writer io.Writer) (_ io.Writer, err error) {
+				w.uncompressedMeter, err = size.NewMeterWithBackupFile(ctx, clk, logger, writer, backupPath, cfg.Statistics.DiskSyncInterval)
+				return w.uncompressedMeter, err
+			})
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// Size of the compressed and uncompressed data is same
+			w.uncompressedMeter = w.compressedMeter
+		}
+	}
+
+	// Create syncer to trigger sync based on counter and meters from the previous steps
+	{
+		w.syncer = syncerFactory(ctx, logger, clk, slice.LocalStorage.DiskSync, w.chain, w)
+	}
+
+	// Create file format writer.
+	// It is entrypoint of the writers chain.
+	{
+		w.formatWriter, err = formatWriterFactory(cfg, w.chain, w.slice)
+		if err != nil {
+			return nil, err
+		}
+
+		// Flush/Close the file format writer at first
+		w.chain.PrependFlusherCloser(w.formatWriter)
+	}
+
+	// Dispatch "open" event
+	if err = volumeEvents.dispatchOnWriterOpen(w); err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+func (w *writer) WriteRecord(timestamp time.Time, values []any) error {
+	// Block Close method
+	w.writeWg.Add(1)
+	defer w.writeWg.Done()
+
+	// Check if the writer is closed
+	if w.isClosed() {
+		return errors.New(`writer is closed`)
+	}
+
+	// Check values count
+	if len(values) != len(w.slice.Columns) {
+		return errors.Errorf(`expected %d columns in the row, given %d`, len(w.slice.Columns), len(values))
+	}
+
+	// Format and write table row
+	if err := w.formatWriter.WriteRecord(values); err != nil {
+		return err
+	}
+
+	notifier := w.syncer.Notifier()
+
+	// Increments number of high-level writes in progress
+	w.acceptedWrites.Add(timestamp, 1)
+
+	// Wait for sync and return sync error, if any
+	if err := notifier.Wait(); err != nil {
+		return err
+	}
```

Comment:
If enabled, the `WriteRecord` method waits until the record is written to the disk, so we provide strong guarantees that data will not be lost. It would be even better if we also had replication on several disks.